src/tri.rs

```diff
@@ -0,0 +1,288 @@
+    D::Smaller: Copy,
+{
+    /// Upper triangular of an array.
+    ///
+    /// Return a copy of the array with elements below the *k*-th diagonal zeroed.
+    /// For arrays with `ndim` exceeding 2, `triu` will apply to the final two axes.
+    /// For 0D and 1D arrays, `triu` will return an unchanged clone.
+    ///
+    /// See also [`ArrayBase::tril`]
+    ///
+    /// ```
+    /// use ndarray::array;
+    ///
+    /// let arr = array![[1, 2, 3], [4, 5, 6], [7, 8, 9]];
+    /// let res = arr.triu(0);
+    /// assert_eq!(res, array![[1, 2, 3], [0, 5, 6], [0, 0, 9]]);
+    /// ```
+    pub fn triu(&self, k: isize) -> Array<A, D>
+    {
+        match self.ndim() > 1 && is_layout_f(&self.dim, &self.strides) {
+            true => {
+                let n = self.ndim();
+                let mut x = self.view();
+                x.swap_axes(n - 2, n - 1);
+                let mut tril = x.tril(-k);
+                tril.swap_axes(n - 2, n - 1);
+
+                tril
+            }
+            false => {
+                let mut res = Array::zeros(self.raw_dim());
+                Zip::indexed(self.rows())
+                    .and(res.rows_mut())
+                    .for_each(|i, src, mut dst| {
+                        let row_num = i.into_dimension().last_elem();
+                        let lower = max(row_num as isize + k, 0);
+                        dst.slice_mut(s![lower..]).assign(&src.slice(s![lower..]));
+                    });
+
+                res
+            }
+        }
+    }
+
+    /// Lower triangular of an array.
+    ///
+    /// Return a copy of the array with elements above the *k*-th diagonal zeroed.
+    /// For arrays with `ndim` exceeding 2, `tril` will apply to the final two axes.
+    /// For 0D and 1D arrays, `tril` will return an unchanged clone.
+    ///
+    /// See also [`ArrayBase::triu`]
+    ///
+    /// ```
+    /// use ndarray::array;
+    ///
+    /// let arr = array![[1, 2, 3], [4, 5, 6], [7, 8, 9]];
+    /// let res = arr.tril(0);
+    /// assert_eq!(res, array![[1, 0, 0], [4, 5, 0], [7, 8, 9]]);
+    /// ```
+    pub fn tril(&self, k: isize) -> Array<A, D>
+    {
+        match self.ndim() > 1 && is_layout_f(&self.dim, &self.strides) {
+            true => {
+                let n = self.ndim();
+                let mut x = self.view();
+                x.swap_axes(n - 2, n - 1);
+                let mut tril = x.triu(-k);
+                tril.swap_axes(n - 2, n - 1);
+
+                tril
+            }
+            false => {
+                let mut res = Array::zeros(self.raw_dim());
+                Zip::indexed(self.rows())
+                    .and(res.rows_mut())
+                    .for_each(|i, src, mut dst| {
+                        // This ncols must go inside the loop to avoid panic on 1D arrays.
+                        // Statistically-neglible difference in performance vs defining ncols at top.
+                        let ncols = src.len_of(Axis(src.ndim() - 1)) as isize;
```

Comment:
I wonder, isn't it simpler to start with

    if self.ndim() <= 1
        return clone or to_owned

remove the first condition in the `match` and put the `ncols` line before the `Zip`? 0D and 1D are simply copied anyway. This would make that rules clear.

Both ways are fine, it's only a suggestion.